### PR TITLE
Feature/spaces administration: display the list of spaces

### PR DESCRIPTION
**Pull request by seifeddineDaoud**

### CODE REVIEW

File: webapp/vue-portlet/pom.xml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        </configuration>
+        <executions>
+          <execution>
+            <id>install node and npm</id>
+            <goals>
+              <goal>install-node-and-npm</goal>
+            </goals>
+            <configuration>
+              <nodeVersion>v10.4.1</nodeVersion>
+              <npmVersion>6.1.0</npmVersion>
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/pom.xml]: Versions should be defined in properties.

File: webapp/vue-portlet/pom.xml [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+          </execution>
+        </executions>
+      </plugin>
+    </plugins>
+  </build>
+  <dependencies>
+    <dependency>
+      <groupId>javax.portlet</groupId>
+      <artifactId>portlet-api</artifactId>
+      <version>2.0</version>
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/pom.xml]: I don't think we need the version here, it is inherited from the parent pom.

File: webapp/vue-portlet/src/main/webapp/META-INF/exo-conf/configuration.xml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/META-INF/exo-conf/configuration.xml]: No need to add this file. Instead declare the new webapp in the global portal container config here : https://github.com/exodev/platform/blob/develop/extension/config/src/main/resources/conf/platform/configuration.xml

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationManageSpaces.vue [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    <div class="uiSearchInput">
+      <input v-model="search" class="showInputSearch" type="text" placeholder="search here"/>
+      <a data-original-title="Search" class="advancedSearch" rel="tooltip" data-placement="bottom" href="#">
+        <i class="uiIconPLF24x24Search" @click="searchSpaces()"></i>
+      </a>
+    </div>
+    <table class="uiGrid table table-hover table-striped">
+      <tr>          
+        <th>
+          Space
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationManageSpaces.vue]: All labels must be internationalized.

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpaces.vue [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+<script>
+import ExoSpacesAdministrationManageSpaces from './ExoSpacesAdministrationManageSpaces.vue';
+import ExoSpacesAdministrationSpacesPermissions from './ExoSpacesAdministrationSpacesPermissions.vue';
+
+export default {
+  components: {
+    'exo-spaces-administration-manage-spaces': ExoSpacesAdministrationManageSpaces,
+    'exo-spaces-administration-manage-permissions' : ExoSpacesAdministrationSpacesPermissions
+  },
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpaces.vue]: In order to make them overridable, all components must be declared globally, like in chat application : https://github.com/exo-addons/chat-application/blob/develop/application/src/main/webapp/vue-app/components/initComponents.js

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/main.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// should expose the locale ressources as REST API 
+const url = '/social-vue-portlet/spaces-administration-app/locale_en.json';
+
+// getting locale ressources
+exoi18n.loadLanguageAsync(lang, url).then(i18n => {
+
+// init Vue app when locale ressources are ready
+  new Vue({
+    el: '#spacesAdministration',
+    render: h => h(ExoSpacesAdministrationSpaces),
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/main.js]: The template of the app must be defined as in the chat application : https://github.com/exo-addons/chat-application/blob/develop/application/src/main/webapp/vue-app/main.js#L52

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationServices.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import {spaceConstants} from './spaceAdministrationConstants.js';
+
+export function getSpaces(){
+  return fetch(`/portal/rest/v1/social/spaces?limit=${spaceConstants.SPACES_PER_PAGE}&returnSize=true`, {credentials: 'include'}).then(resp => resp.json());
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationServices.js]: REST base url should be defined in constants (like SOCIAL_USER_API).

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+        } else {
+          item.avatarUrl = '/eXoSkin/skin/images/system/SpaceAvtDefault.png';
+        }
+      }
+      return `
+        <div class="item"> <img class="avatarMini" src="${item.avatarUrl}"> ${escape(item.value)}</div>
+      `;
+    },
+    addSuggestedItem(item) {
+      if($('#add-user-suggestor') && $('#add-user-suggestor').length && $('#add-user-suggestor')[0].selectize) {
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue]: Do we really have to use jquery here ? Cannot we use Vue data binding instead ?

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationServices.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+export function getSpaceLinkSetting(spaceDisplayName){
+  const nameSpace = spaceDisplayName.toLowerCase().split(' ').join('_');
+  return `${spaceConstants.PORTAL}${spaceConstants.PROFILE_SPACE_LINK}${nameSpace}/${nameSpace}/settings`;
+}
+
+export function getAvatar(user) {
+  return `${spaceConstants.SOCIAL_USER_API}${user}/avatar`;
+}
+
+export function getGuests(query) {
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationServices.js]: Why do we use the term "guests" ? We should call this method findUsers IMO (and findGroups for the groups).

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+          }
+        }      
+      }
+    },
+    findGuests (query, callback) {
+      if (!query.length) {
+        return callback(); 
+      }
+      spaceAdministrationServices.getGuests(query).then(data => {
+        spaceAdministrationServices.getGuestsGroups(query).then(group => {
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue]: These 2 methods should be called in parallel. You can use Promise.all() for that (here is an example : https://github.com/exodev/platform/blob/develop/extension/webapp/src/main/webapp/javascript/document-preview.js#L80).

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue [hunk 11]
```diff
@@ -0,0 +1,305 @@
...
+          },
+          onItemAdd(item) {
+            component.addSuggestedItem(item);
+          },
+          onItemRemove(item) {
+            component.removeSuggestedItem(item);
+          },
+          sortField: [{field: 'order'}, {field: '$score'}],
+          providers: {
+            'exo:social': component.findGuests
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue]: "exo:social" should be changed to "exo:spacesAdministration".

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue [hunk 12]
```diff
@@ -0,0 +1,305 @@
...
+          this.permissionManagers = data.value;
+        }
+        this.initSuggesterManageSpace();
+      });
+    },
+    initSuggester() {
+      const $guestsFormsSuggestor = $('#add-user-suggestor');
+      if($guestsFormsSuggestor && $guestsFormsSuggestor.length && $guestsFormsSuggestor.suggester) {
+        const component = this;
+        const suggesterData = {
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/components/ExoSpacesAdministrationSpacesPermissions.vue]: Cannot we factorize both suggesterData objects ?

File: webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationConstants.js [hunk 13]
```diff
@@ -0,0 +1,9 @@
+export const spaceConstants = {
+  ENV: eXo.env.portal || '',
+  PORTAL: eXo.env.portal.context || '',
+  PORTAL_NAME: eXo.env.portal.portalName || '',
+  PROFILE_SPACE_LINK: '/g/:spaces:',
+  SOCIAL_USER_API: `${eXo.env.portal.context}/${eXo.env.portal.rest}/v1/social/users/`,
+  SPACES_PER_PAGE: 2,
```
Comment (thomasdelhomenie) [inline webapp/vue-portlet/src/main/webapp/spaces-administration-app/spaceAdministrationConstants.js]: Must be changed to 30 (I know it was set to 2 for testing purpose).

File: component/service/src/main/java/org/exoplatform/social/rest/impl/space/SpaceRestResourcesV1.java [hunk 14]
```diff
@@ -109,31 +116,57 @@ public Response getSpaces(@Context UriInfo uriInfo,
...
+    if (StringUtils.isNotBlank(sort)) {
+      try {
+        SortBy sortBy = Sorting.SortBy.valueOf(sort.toUpperCase());
+        OrderBy orderBy = Sorting.OrderBy.ASC;
+        if (StringUtils.isNotBlank(order)) {
+          orderBy = Sorting.OrderBy.valueOf(order.toUpperCase());
+        }
+        spaceFilter.setSorting(new Sorting(sortBy, orderBy));
+      } catch(Exception e) {
+
```
[the remark this answers is not in the document]
Comment (seifeddineDaoud) [reply component/service/src/main/java/org/exoplatform/social/rest/impl/space/SpaceRestResourcesV1.java]: No , we don't need the try/catch . I removed it.

File: component/service/src/test/java/org/exoplatform/social/rest/impl/spacesadministration/SpacesAdministrationRestResourcesTest.java [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+    startSessionAs("mary");
+    
+    // When
+    ContainerResponse response = service("GET", getURLResource("spacesAdministration/groups"), "", null, null, "mary");
+    
+    // Then
+    assertNotNull(response);
+    assertEquals(401, response.getStatus());
+  }
+  */
```
[the remark this answers is not in the document]
Comment (seifeddineDaoud) [reply component/service/src/test/java/org/exoplatform/social/rest/impl/spacesadministration/SpacesAdministrationRestResourcesTest.java]: Done

File: extension/war/src/main/webapp/groovy/social/webui/space/UISpaceSearch.gtmpl [hunk 16]
```diff
@@ -3,7 +3,7 @@
 def rcontext = _ctx.getRequestContext() ;
 def typeOfRelation = uicomponent.getTypeOfRelation();
 def spaceURL = uicomponent.getSpaceURL();
-    
+def checkPermission = uicomponent.checkPermissionCreateSpace();
```
[the remark this answers is not in the document]
Comment (seifeddineDaoud) [reply extension/war/src/main/webapp/groovy/social/webui/space/UISpaceSearch.gtmpl]: OK Done

File: webapp/vue-portlet/src/main/webapp/WEB-INF/conf/configuration.xml.ori [hunk 17]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (seifeddineDaoud) [reply webapp/vue-portlet/src/main/webapp/WEB-INF/conf/configuration.xml.ori]: Done

**Pull request merged**